### PR TITLE
chore: clap v3

### LINKED ISSUE #8475: Research upgrade to clap v3
Deno uses `clap` as its CLI parser. It looks like clap v3 is behind the corner so I think it's high time to do some research what it would take to upgrade our `cli/flags.rs`.

Keep in mind, [`clap` authors recommended](https://docs.rs/clap/3.0.0-beta.2/clap/) to stay on `v2.33` still as the API is not fully stable. So I'd be happy to have a PR open until stable v3 is released.

### LINKED ISSUE #12118: `NO_COLOR=1` does not works for "Did you mean ..."
It seems `NO_COLOR=1` does not affect coloring on "Did you mean ..." part.

CleanShot 2021-09-17 at 15 52 20

My Deno version is 

```
deno 1.14.0 (release, x86_64-apple-darwin)
v8 9.4.146.15
typescript 4.4.2
```

Thanks for the project 👍

### PR DESCRIPTION
**Pull request by crowlKats**
Closes #8475
Closes #12118

### CODE REVIEW

**Comment by lucacasonato** [comment]
Also closes #12118

File: Cargo.lock [hunk 1]
```diff
@@ -71,15 +71,6 @@ dependencies = [
...
 ]
 
-[[package]]
-name = "ansi_term"
-version = "0.11.0"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "ee49baf6cb617b853aa8d93bf420db2383fab46d314482ca2803b40d5fde979b"
-dependencies = [
- "winapi 0.3.9",
-]
```
Comment (ry) [inline Cargo.lock:81]: Nice!

**Review by ry — APPROVED** [review]
LGTM

commit message should probably be "upgrade: clap v3"

**Comment by crowlKats** [comment]
DO NOT MERGE. discussed with @bartlomieju and we came to the conclusion that this PR should wait for 2.0, as there are some fundamental changes to clap's parsing, which would likely be breaking changes in various cases

**Comment by ry** [comment]
@crowlKats Can you give an example of what changed?

**Comment by crowlKats** [comment]
I have tried to keep behaviour same as current, but i cant be certain that it actually fully is. one test was failing with a weird case and had to change what we were doing, but that change maybe would need to be applied to other subcommands, or maybe not. its a bit hairy. also, cargo is being wary of updating as well https://github.com/rust-lang/cargo/pull/10253#issuecomment-1005045779 (and subsequent comments)

**Comment by ry** [comment]
@crowlKats and I have done a little experimentation and decided that if this does introduce bugs, they are very not common, it works in all the instances we tried. I suggest landing.

LGTM

**Review by bartlomieju — APPROVED** [review]

**Pull request merged**